### PR TITLE
Initializes Slint gui

### LINKED ISSUE #238: Try Slint as a new GUI
Now they just released a new version with a new license that is compatible with MIT license. So it is a good opportunity to try it to see if we can get rid of C++.

The new GUI should be placed in `src/gui` while we are experimenting with it. It should not ship with the release to the end users until it comes out from experimental.

### PR DESCRIPTION
**Pull request by SuchAFuriousDeath**
Closes #238.

### CODE REVIEW

**Comment by ultimaweapon** [comment]
Require #1083 to merge first.

**Comment by ultimaweapon** [comment]
You need to change `add_crate(gui)` to `add_crate(gui LIBRARY)`.

**Comment by ultimaweapon** [comment]
#238 should be able to close by this PR.

**Comment by ultimaweapon** [comment]
Seems like this PR is going too big. My idea is just pushing only the GUI without functionality first.

File: gui/build.rs [hunk 1]
```diff
@@ -8,6 +9,30 @@ const LINUX_INCLUDE: &str = r#"
...
+    )
+    .unwrap();
+
+    // Compile main
+    slint_build::compile_with_config(
+        "slint/main.slint",
+        CompilerConfiguration::new().with_style(String::from("fluent-dark")),
+    )
+    .unwrap();
+}
```
Comment (ultimaweapon) [inline gui/build.rs:33]: Is it possible to use a wildcard here? Also I think it is better to rename the folder to the other name (e.g. `ui`) so we don't need to use `::slint` to reference the Slint crate.

File: gui/build.rs [hunk 2]
```diff
@@ -8,6 +9,30 @@ const LINUX_INCLUDE: &str = r#"
 "#;
 
 fn main() {
+    if std::env::var("CARGO_FEATURE_GUI_SLINT").is_ok() {
```
Comment (ultimaweapon) [inline gui/build.rs:12]: There should be a better way than this. Feature flag should be used to enable additional code instead of using like this.

**Review by ultimaweapon — CHANGES_REQUESTED** [review]

File: gui/build.rs (hunk 2 above)
Comment (SuchAFuriousDeath) [reply gui/build.rs:12]: I tried, but AFAIK:
1. there can be only one build script per crate.
2. There is no way to tell if we are compiling the library or the binary.

File: gui/build.rs (hunk 1 above)
Comment (SuchAFuriousDeath) [reply gui/build.rs:33]: I renamed the module, but wildcards don't seems possible based on the the code: Maybe with some compile time magic

Comment (ultimaweapon) [reply gui/build.rs:33]: In that case we can recursive list all files in the directory.

File: gui/build.rs (hunk 2 above)
Comment (ultimaweapon) [reply gui/build.rs:12]: Have you try `CARGO_BIN_NAME`?

Comment (SuchAFuriousDeath) [reply gui/build.rs:12]: oh :no_mouth:

Comment (SuchAFuriousDeath) [reply gui/build.rs:12]: I tried CARGO_BIN_PATH, but it's not set :(.

Comment (ultimaweapon) [reply gui/build.rs:12]: Does Slint support including those files in Rust instead of building it here? Something like `include_slint!("file.slint").

Comment (SuchAFuriousDeath) [reply gui/build.rs:12]: I don't think so. The only other option is including slint source directly in Rust, there's a macro for that

Comment (ultimaweapon) [reply gui/build.rs:12]: Another idea is putting the FFI for Qt behind a feature flag and enabling it from CMake.

Comment (SuchAFuriousDeath) [reply gui/build.rs:12]: I did that, take a look

File: gui/build.rs (hunk 1 above)
Comment (SuchAFuriousDeath) [reply gui/build.rs]: I removed the other file. It seems you can have only one slint file (You can export multiple things, but only from one file)

File: gui/build.rs [hunk 3]
```diff
@@ -8,6 +9,23 @@ const LINUX_INCLUDE: &str = r#"
 "#;
 
 fn main() {
+    if std::env::var("CARGO_FEATURE_GUI_SLINT").is_ok_and(|var| var == "1") {
+        build_bin();
+    } else if std::env::var("CARGO_FEATURE_QT_FFI").is_ok_and(|var| var == "1") {
```
Comment (ultimaweapon) [inline gui/build.rs:14]: This should not be `else if`. See https://doc.rust-lang.org/cargo/reference/features.html#feature-unification for more information about this.

File: gui/build.rs [hunk 4]
```diff
@@ -8,6 +9,23 @@ const LINUX_INCLUDE: &str = r#"
...
+    }
+}
+
+fn build_bin() {
+    // Compile main
+    slint_build::compile_with_config(
+        "slint/main.slint",
+        CompilerConfiguration::new().with_style(String::from("fluent-dark")),
+    )
+    .unwrap();
```
Comment (ultimaweapon) [inline gui/build.rs:25]: Better to recursive enumerate all files in this directory and compile all of `.slint` files.

File: gui/Cargo.toml [hunk 5]
```diff
@@ -5,6 +5,19 @@ edition = "2021"
...
 crate-type = ["staticlib"]
+required-features = ["qt_ffi"]
+
+[[bin]]
+name = "gui"
+path = "src/main.rs"
+required-features = ["gui_slint"]
+
+[features]
+gui_slint = [
```
Comment (ultimaweapon) [inline gui/Cargo.toml:16]: I think the name is too long. Just `slint` should be enough.

File: gui/Cargo.toml [hunk 6]
```diff
@@ -5,6 +5,19 @@ edition = "2021"
...
+    "slint",
+    "clap"
+]
+qt_ffi = []
```
Comment (ultimaweapon) [inline gui/Cargo.toml:20]: Same here, just `qt` should be enough.

File: gui/Cargo.toml [hunk 7]
```diff
@@ -5,6 +5,19 @@ edition = "2021"
 
 [lib]
 crate-type = ["staticlib"]
+required-features = ["qt_ffi"]
+
+[[bin]]
+name = "gui"
```
Comment (ultimaweapon) [inline gui/Cargo.toml:11]: This will be the name of output file so maybe `obliteration` should be better?

File: gui/Cargo.toml [hunk 8]
```diff
@@ -23,9 +36,18 @@ uuid = { version = "1.10.0", features = ["serde", "v4"] }
 
 [target.'cfg(target_arch = "x86_64")'.dependencies]
 x86-64 = { path = "../arch/x86-64" }
+raw-window-handle = "0.6"
```
Comment (ultimaweapon) [inline gui/Cargo.toml:39]: Is not this should be optional?

File: gui/Cargo.toml [hunk 9]
```diff
@@ -23,9 +36,18 @@ uuid = { version = "1.10.0", features = ["serde", "v4"] }
...
+clap = { version = "4.5.20", features = ["derive"], optional = true }
+
+[dependencies.slint]
+git = "https://github.com/slint-ui/slint"
+rev = "875ca075fb5b2dfe4c3ab0a499d5759412fc1395"
```
Comment (ultimaweapon) [inline gui/Cargo.toml:44]: Is it possible to use the published version here? The same reason as `cbindgen`.

File: gui/Cargo.toml [hunk 10]
```diff
@@ -46,3 +68,4 @@ objc = "0.2.7"
 
 [build-dependencies]
 cbindgen = "0.26.0"
+slint-build = { git = "https://github.com/slint-ui/slint", rev = "875ca075fb5b2dfe4c3ab0a499d5759412fc1395" }
```
Comment (ultimaweapon) [inline gui/Cargo.toml:71]: Same here.

File: cmake/cargo.cmake [hunk 11]
```diff
@@ -157,6 +157,10 @@ function(add_crate crate)
     list(APPEND build_args "$<IF:$<CONFIG:Debug>,--profile=dev,--release>")
     list(APPEND build_args ${arg_ARGS})
 
+    if(arg_LIBRARY)
+        list(APPEND build_args "--lib")
+    endif()
+
```
Comment (ultimaweapon) [inline cmake/cargo.cmake:163]: This one should be inside `staticlib` block below so it works with a single type crate.

**Review by ultimaweapon — CHANGES_REQUESTED** [review]

File: gui/build.rs (hunk 4 above)
Comment (SuchAFuriousDeath) [reply gui/build.rs:27]: We will only be compiling one file. Slint doesn't support multiple files.

File: gui/Cargo.toml (hunk 9 above)
Comment (SuchAFuriousDeath) [reply gui/Cargo.toml:43]: Not at the moment

File: gui/Cargo.toml (hunk 10 above)
Comment (SuchAFuriousDeath) [reply gui/Cargo.toml:72]: Same here too :(

File: gui/Cargo.toml (hunk 9 above)
Comment (ultimaweapon) [reply gui/Cargo.toml:45]: In that case we should create an issue for this so when the release we need available we can fix this.

File: cmake/cargo.cmake [hunk 12]
```diff
@@ -178,6 +178,11 @@ function(add_crate crate)
 
         if(${kind} STREQUAL "staticlib")
             add_library(${crate} STATIC IMPORTED)
+
+            if(arg_LIBRARY)
```
Comment (ultimaweapon) [inline cmake/cargo.cmake:182]: We don't need this `if` since we are in the `staticlib` block.

File: gui/src/screen/vulkan/mod.rs [hunk 13]
```diff
@@ -31,77 +19,44 @@ pub struct Vulkan {
...
+        let appinfo = ApplicationInfo::default()
+            .application_name(c"Obliteration")
+            .application_version(0)
+            .api_version(ash::vk::make_api_version(0, 1, 0, 0));
+
+        let create_info = InstanceCreateInfo::default().application_info(&appinfo);
+
+        let entry = ash::Entry::linked();
+
+        let instance = unsafe { entry.create_instance(&create_info, None) }
```
Comment (ultimaweapon) [inline gui/src/screen/vulkan/mod.rs:33]: This should reuse the same instance from the Qt side.

**Review by ultimaweapon — CHANGES_REQUESTED** [review]

**Comment by VocalFan** [comment]
I will just keep watching from the distance... Also known as emails from github.

File: gui/src/rlim.rs [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+                    if unsafe { libc::setrlimit(libc::RLIMIT_NOFILE, &rlim) } != 0 {
+                        eprintln!(
+                            "Failed to set RLIMIT_NOFILE: {}",
+                            std::io::Error::last_os_error()
+                        );
+                    }
+                }
+            }
+            _ => {
+                eprintln!(
```
Comment (ultimaweapon) [inline gui/src/rlim.rs:25]: This should be a message box otherwise the user won't know what is going wrong.

File: gui/src/rlim.rs [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+
+        match ret {
+            0 => {
+                let mut rlim = unsafe { rlim.assume_init() };
+
+                if rlim.rlim_cur < rlim.rlim_max {
+                    rlim.rlim_cur = rlim.rlim_max;
+
+                    if unsafe { libc::setrlimit(libc::RLIMIT_NOFILE, &rlim) } != 0 {
+                        eprintln!(
```
Comment (ultimaweapon) [inline gui/src/rlim.rs:17]: Same here.

**Review by ultimaweapon — CHANGES_REQUESTED** [review]

File: gui/src/rlim.rs [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+
+            Ok(())
+        }
+        _ => Err(RlimitError::GetRlimitFailed(std::io::Error::last_os_error())),
+    }
+}
+
+#[derive(Debug, Error)]
+pub(crate) enum RlimitError {
+    #[error("failed to get RLIMIT_NOFILE -> {0}")]
```
Comment (ultimaweapon) [inline gui/src/rlim.rs:29]: Printing the nested error here will make it impossible for the caller to format the error chain. Better to use `Error::source` on the caller side instead.

File: gui/src/rlim.rs [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+    GetRlimitFailed(#[source] std::io::Error),
+
+    #[error("failed to set RLIMIT_NOFILE -> {0}")]
```
Comment (ultimaweapon) [inline gui/src/rlim.rs:32]: Same here.

**Review by ultimaweapon — CHANGES_REQUESTED** [review]

File: gui/src/vmm/hv/mod.rs [hunk 18]
```diff
@@ -24,6 +24,15 @@ pub type Default = self::os::Hvf;
...
 
+#[cfg(target_os = "linux")]
+pub type DefaultError = self::os::KvmError;
+
+#[cfg(target_os = "macos")]
+pub type DefaultError = self::os::HvfError;
+
+#[cfg(target_os = "windows")]
+pub type DefaultError = self::os::WhpError;
+
```
Comment (ultimaweapon) [inline gui/src/vmm/hv/mod.rs:35]: These should be `HypervisorError`.

**Review by ultimaweapon — CHANGES_REQUESTED** [review]

File: gui/src/vmm/mod.rs [hunk 19]
```diff
@@ -684,6 +444,116 @@ pub enum DebugResult {
     Error { reason: *mut RustError },
 }
 
+#[derive(Debug, Error)]
+pub enum StartVmmError {
```
Comment (ultimaweapon) [inline gui/src/vmm/mod.rs:448]: Now `VmmError` has been removed so we can reuse this name here.

**Review by ultimaweapon — CHANGES_REQUESTED** [review]

File: gui/src/main.rs [hunk 20]
```diff
@@ -0,0 +1,165 @@
...
+    Ok,
+    Err(ApplicationError),
+}
+
+impl Termination for AppExit {
+    fn report(self) -> ExitCode {
+        match self {
+            AppExit::Ok => ExitCode::SUCCESS,
+            AppExit::Err(e) => {
+                eprintln!("{}", e);
```
Comment (ultimaweapon) [inline gui/src/main.rs:130]: We should use a message box instead otherwise it won't be seen by the user.

File: gui/src/main.rs [hunk 21]
```diff
@@ -0,0 +1,175 @@
...
+
+impl Termination for AppExit {
+    fn report(self) -> ExitCode {
+        match self {
+            AppExit::Ok => ExitCode::SUCCESS,
+            AppExit::Err(e) => {
+                let _ = ui::ErrorDialog::new()
+                    .and_then(|error_dialog| {
+                        error_dialog.set_message(SharedString::from(format!(
+                            "Error setting rlimit: {}",
```
Comment (ultimaweapon) [inline gui/src/main.rs:133]: Seems like the wrong message here.

File: gui/src/main.rs [hunk 22]
```diff
@@ -0,0 +1,175 @@
...
+                            full_error_reason(e)
+                        )));
+
+                        error_dialog.run()
+                    })
+                    .inspect_err(|e| eprintln!("Error displaying error dialog: {e}"));
```
Comment (ultimaweapon) [inline gui/src/main.rs:139]: I think unwrap should be better because this should never fail.

File: gui/src/main.rs [hunk 23]
```diff
@@ -0,0 +1,175 @@
...
+        let _ = ui::ErrorDialog::new()
+            .and_then(|error_dialog| {
+                error_dialog.set_message(SharedString::from(format!(
+                    "Error setting rlimit: {}",
+                    full_error_reason(e)
+                )));
+
+                error_dialog.run()
+            })
+            .inspect_err(|e| eprintln!("Error displaying error dialog: {e}"));
```
Comment (ultimaweapon) [inline gui/src/main.rs:40]: Same here.

**Review by ultimaweapon — CHANGES_REQUESTED** [review]

**Review by ultimaweapon — APPROVED** [review]

**Pull request merged**